### DISCUSSION TITLE
Modelling a displacement controlled RVE

### DISCUSSION THREAD

**Discussion by Flolaffel** [idaholab/moose#27858] [Q&A Modules: Solid mechanics]
Hello,

I've been trying to model a displacement controlled RVE simulation in MOOSE for a while now but didn't really achieve a satisfying result yet.

One approach I got to work is using `LinearNodalConstraint` to impose two constraints (disp_x and disp_y) per node pair on opposing side sets.

```
nx = 3
ny = 3

[Mesh]
  [domain]
    type = GeneratedMeshGenerator
    dim = 2
    xmin = 0
    xmax = 3
    ymin = 0
    ymax = 3
    nx = ${nx}
    ny = ${ny}
  []
  [top_right]
    type = ExtraNodesetGenerator
    input = domain
    new_boundary = top_right
    coord = '${fparse nx} ${fparse ny} 0'
  []
  [bottom_right]
    type = ExtraNodesetGenerator
    input = top_right
    new_boundary = bottom_right
    coord = '${fparse nx} 0 0'
  []
  [top_left]
    type = ExtraNodesetGenerator
    input = bottom_right
    new_boundary = top_left
    coord = '0 ${fparse ny} 0'
  []
  [bottom_left]
    type = ExtraNodesetGenerator
    input = top_left
    new_boundary = bottom_left
    coord = '0 0 0'
  []
[]

[GlobalParams]
  displacements = 'disp_x disp_y'
[]

[Modules/TensorMechanics/Master]
  [all]
    strain = SMALL
    add_variables = true
    incremental = false
    generate_output = 'vonmises_stress'
  []
[]

[BCs]
  [no_y]
    type = DirichletBC
    variable = disp_y
    boundary = 'bottom_left bottom_right top_left'
    value = 0
  []
  [no_x]
    type = DirichletBC
    variable = disp_x
    boundary = 'bottom_left'
    value = 0
  []
  [disp]
    type = DirichletBC
    variable = disp_x
    boundary = 'top_left'
    value = 0.1
  []
[]

[Constraints]
  [lr1_x]
    type = LinearNodalConstraint
    variable = disp_x
    primary = '3 6 0'
    weights = '1 1 -1'
    secondary_node_ids = 7
    penalty = 1e6
    formulation = kinematic
  []
  [lr1_y]
    type = LinearNodalConstraint
    variable = disp_y
    primary = '3 6 0'
    weights = '1 1 -1'
    secondary_node_ids = 7
    penalty = 1e6
    formulation = kinematic
  []
  [lr2_x]
    type = LinearNodalConstraint
    variable = disp_x
    primary = '9 6 0'
    weights = '1 1 -1'
    secondary_node_ids = 11
    penalty = 1e6
    formulation = kinematic
  []
  [lr2_y]
    type = LinearNodalConstraint
    variable = disp_y
    primary = '9 6 0'
    weights = '1 1 -1'
    secondary_node_ids = 11
    penalty = 1e6
    formulation = kinematic
  []
  [lr3_x]
    type = LinearNodalConstraint
    variable = disp_x
    primary = '13 6 0'
    weights = '1 1 -1'
    secondary_node_ids = 15
    penalty = 1e6
    formulation = kinematic
  []
  [lr3_y]
    type = LinearNodalConstraint
    variable = disp_y
    primary = '13 6 0'
    weights = '1 1 -1'
    secondary_node_ids = 15
    penalty = 1e6
    formulation = kinematic
  []
  [tb1_x]
    type = LinearNodalConstraint
    variable = disp_x
    primary = '1 13 0'
    weights = '1 1 -1'
    secondary_node_ids = 12
    penalty = 1e6
    formulation = kinematic
  []
  [tb1_y]
    type = LinearNodalConstraint
    variable = disp_y
    primary = '1 13 0'
    weights = '1 1 -1'
    secondary_node_ids = 12
    penalty = 1e6
    formulation = kinematic
  []
  [tb2_x]
    type = LinearNodalConstraint
    variable = disp_x
    primary = '4 13 0'
    weights = '1 1 -1'
    secondary_node_ids = 14
    penalty = 1e6
    formulation = kinematic
  []
  [tb2_y]
    type = LinearNodalConstraint
    variable = disp_y
    primary = '4 13 0'
    weights = '1 1 -1'
    secondary_node_ids = 12
    penalty = 1e6
    formulation = kinematic
  []
[]

[Materials]
  [elasticity]
    type = ComputeIsotropicElasticityTensor
    youngs_modulus = 210000
    poissons_ratio = 0.3
  []
  [stress]
    type = ComputeLinearElasticStress
  []
[]

[Preconditioning]
  [smp]
    type = SMP
    full = true
  []
[]

[Executioner]
  type = Steady
  solve_type = 'PJFNK'
  petsc_options_iname = '-pc_type -pc_factor_shift_type -pc_factor_shift_amount'
  petsc_options_value = 'lu       NONZERO               1e-15'
  nl_abs_tol = 1e-8
  nl_rel_tol = 1e-12
[]

[Outputs]
  exodus = true
[]
```

The result is what I'm aiming for, so we can use that as a reference. The problem with this approach however is the convergence. I tried various PETsc settings, preconditioning and solve type combinations. This here is the best I could do which unfortunately is not good enough for the actual application I have in mind (convergence fails there).

My preferred approach would be just using the `AddPeriodicBCAction` like this:

```
[BCs]
  [Periodic]
    [x]
      variable = disp_x
      auto_direction = 'x y'
    []
    [y]
      variable = disp_y
      auto_direction = 'x y'
    []
  []
[]
```

Unfortunately replacing the `LinearNodalConstraints` with this snippet doesn't have the desired result but rather delivers a non-displaced RVE.

So my questions are:
1. Am I missing something with the `AddPeriodicBCAction` that keeps it from working?
2. Is there another preferred approach for this type of problem that I'm missing? Keep in mind I need the simulation to be displacement controlled. So a global strain approach would not be possible since we apply a stress tensor instead of a deformation gradient, would it?
3. If there is no other way, how could I improve the convergence on my `LinearNodalConstraint` approach?

**Comment by GiudGiud** [comment]
Hello

`AddPeriodicBCAction` can be pretty finicky in auto-mode as iirc it does not report when it does not figure out how to set up the periodicity. Did you try the manual mode?
And did you check that output follows the periodicity as expected? Maybe use a simple diffusion problem to do so, instead of displacements

Can you reduce the penalty on those constraints? Does it need to be so high? This affects the conditioning of the matrix usually. you can check the conditioning using the SVD preconditioner, see this page
https://mooseframework.inl.gov/moose/application_usage/failed_solves.html

Does NEWTON method work for you? Is the Jacobian exact as far as you know or is there a known missing term?

**Reply by Flolaffel** [reply]
The detection works fine. The only problem I'm facing is a over constrained model when using the `AddPeriodicBCAction`. I'd like to control the model deformation by only applying BCs to the corner nodes (which is what is happening in my attached input with `LinearNodalConstraint`). This works through coupling the displacement of each secondary node with the displacements of the corresponding corner nodes which is not the same as simply matching the x/y displacement of a node to the corresponding node on the opposite site. I'm guessing that's what the `AddPeriodicBCAction` is trying to do here. If I apply my displacement to a node inside my domain, all the corner nodes act fixed. So applying a displacement there essentially does nothing in this case.

> Is the Jacobian exact as far as you know or is there a known missing term?

AFAIK the standard MOOSE mechanics kernel works with inexact Jacobians. I switched to the Lagrangian kernel system now to make sure everything is correct. But since I'm using small deformations anyway there souldn't be a difference I guess. I don't know how MOOSE handles the Jacobian contribution of the `LinearNodalConstraints`.

> Does NEWTON method work for you?

Newton converges worse than PJFNK, if at all

> you can check the conditioning using the SVD preconditioner, see this page

The SVD reports several singular values. I can't track down the reason for that though. I reduced my input to the bare minimum and still receive singular values. What is happening here?

```
[Mesh]
  [domain]
    type = GeneratedMeshGenerator
    dim = 2
    xmin = 0
    xmax = 3
    ymin = 0
    ymax = 3
    nx = 3
    ny = 3
  []
[]

[GlobalParams]
  displacements = 'disp_x disp_y'
[]

[Modules/TensorMechanics/Master]
  [all]
    strain = SMALL
    add_variables = true
    incremental = false
    new_system = true
    formulation = TOTAL
  []
[]

[BCs]
  [no_y]
    type = DirichletBC
    variable = disp_y
    boundary = 'bottom top'
    value = 0
  []
  [no_x]
    type = DirichletBC
    variable = disp_x
    boundary = 'bottom'
    value = 0
  []
  [disp]
    type = DirichletBC
    variable = disp_x
    boundary = 'top'
    value = 0.5
  []
[]

[Materials]
  [elasticity]
    type = ComputeIsotropicElasticityTensor
    youngs_modulus = 210000
    poissons_ratio = 0.3
  []
  [stress]
    type = ComputeLagrangianLinearElasticStress
  []
[]

[Preconditioning]
  [smp]
    type = SMP
    full = true
  []
[]

[Executioner]
  type = Steady
  solve_type = PJFNK
  petsc_options = '-pc_svd_monitor'
  petsc_options_iname = '-pc_type'
  petsc_options_value = 'svd'
  nl_abs_tol = 1e-8
  nl_rel_tol = 1e-12
[]

[Outputs]
  exodus = true
[]
```

**Reply by GiudGiud** [reply]
> The SVD reports several singular values. I can't track down the reason for that though. I reduced my input to the bare minimum and still receive singular values. What is happening here?

somehow, the problem is not well-posed and can have multiple solutions. 
In mechanics this is usually due to extra rotational or translational degrees of freedom. You can remove these by setting more dirichletBCs on displacements
or by adding constraints (but this tends to hurt convergence too) on the integral values of displacement

**Reply by Flolaffel** [reply]
I have a feeling there may have been a misunderstanding.

The MOOSE documentation says 

> If the svd monitor reports ANY singular value, your problem is ill-posed.

But AFAIK every matrix has singular values (which can be retrieved through SVD) and problems only arise when the singular values are close to 0. So shouldn't the docs be along the lines of 

> If the svd monitor reports ANY singular value **close to 0**, your problem is ill-posed.

My setup doesn't have any singular values close to 0 so I don't think it's ill-posed. Am I wrong?

**Reply by GiudGiud** [reply]
what does the svd monitor report?

**Reply by GiudGiud** [reply]
but yes you are right for the wording. I m not sure why I wrote that wrong

**Reply by Flolaffel** [reply]
SVD Monitor reports something like this:

```
SVD: condition number 5.221156179857e+07, 0 of 32 singular values are (nearly) zero
SVD: smallest singular values: 6.009961766392e-02 1.539427820447e-01 2.672794409233e-01 5.667418564167e-01 7.671714963266e-01
SVD: largest singular values : 1.251704995744e+06 1.800887895077e+06 1.837366659253e+06 3.118666628197e+06 3.137894901730e+06
```

I've also run the Jacobian debug script. I get

```
Kernel for variable 'disp_x':
  (0,0) On-diagonal Jacobian is questionable (off by 10.87 %)
  (0,1) Off-diagonal Jacobian for variable 'disp_y' is questionable (off by 10.56 %)

Kernel for variable 'disp_y':
  (1,0) Off-diagonal Jacobian for variable 'disp_x' is questionable (off by 10.56 %)
  (1,1) On-diagonal Jacobian is questionable (off by 10.87 %)
```

for the input I originally attached to my question. When I comment out the `LinearNodalConstraints` I get

`No errors detected. :-)`

Is that a reason to believe `LinearNodalConstraint` has an incorrect contribution to the Jacobian?

**Reply by GiudGiud** [reply]
Hello
It does seem so. I havent used that object much. @lindsayad may have more insights.
Fixing the Jacobian can improve convergence

No 0 singular value means the problem is well-posed (in a math sense).

**Reply by lindsayad** [reply]
I've never liked the `analyze_jacobian.py` script as much as printing the actual matrix. You can do this with `-snes_test_jacobian -snes_test_jacobian_view`. I can believe that the `LinearNodalConstraint` doesn't have the correct Jacobian especially under mesh deformation (mesh deformation makes it pretty tough to get the Jacobians perfect, even for some AD objects ... need to get that geometric AD proposal awarded @GiudGiud 😄 ). It would be good to verify the Jacobian is right/wrong with a `-snes_test_jacobian` test

As far modeling RVEs in general, I'm going to ping @dschwen @laagesen @sapitts to see if they have any thoughts

**Reply by Flolaffel** [reply]
@lindsayad  Okay so I've run it with `-snes_test_jacobian` which reports

`||J - Jfd||_F/||J||_F = 0.336482, ||J - Jfd||_F = 2.02017e+06`

I then tried switching to `FDP` which does improve convergence but no to the point where I can solve the problem with `NEWTON`. Result:

`||J - Jfd||_F/||J||_F = 0.0905084, ||J - Jfd||_F = 600700.`

I find it surprising that there is a difference between the two FD Jacobians. I guess MOOSE and PETsc slightly differ in the way it's calculated?

**Reply by lindsayad** [reply]
With FDP we use "coloring" based on the mesh connectivity to remove unnecessary differencing operations. With `-snes_test_jacobian` the entire matrix structure is differenced, so it's very slow but there is no possibility for errors. In your case your constraints have no tie to the mesh connectivity (they can be totally across the mesh) so FDP will not work well

**Reply by Flolaffel** [reply]
Okay but since `-snes_test_jacobian` shows significant errors, something is definitely going wrong

**Reply by lindsayad** [reply]
I could have guaranteed that FDP would have a wrong Jacobian before you tried it. We should focus on the difference with SMP. Clearly there is a difference that should be fixed. I just don't think anyone from our team will have the time for it anytime soon. We would certainly welcome an external contribution that fixed it

**Reply by lindsayad** [reply]
I've opened https://github.com/idaholab/moose/issues/27955 to document that we indeed have an issue

**Reply by Flolaffel** [reply]
> We should focus on the difference with SMP

Yes I agree, that's what I meant. Excuse my ambiguity.

> We would certainly welcome an external contribution that fixed it

I will try and have a go at it.

**Reply by Flolaffel** [reply]
@lindsayad Could you please provide a resource for the theoretic basics of the kinematic method of constraint enforcement used in MOOSE? I'm not sure, if I'm familiar with this one.

**Reply by lindsayad** [reply]
@bwspenc or @recuero do you have a reference for kinematic constraint enforcement?

**Reply by Traiwit** [reply]
Hi guys, 

just wondering if we have solved this issue with `Kinematic` method on `LinearNodalConstraint`? 

to me `Penalty` method works fine, but it's much more expensive if I have more than a million constraint pairs

I believe `Kinematic` will be cheaper but it also wont converge for my problem as well

**Reply by GiudGiud** [reply]
@bwspenc @recuero
